### PR TITLE
Trim whitespace from resource values

### PR DESCRIPTION
**Pull request by flyingtoaster0**
Proposed fix for issue #2014

### CODE REVIEW

**Comment by jongerrish** [comment]
I feel like we should be handling this at a lower level, see:-

https://github.com/robolectric/robolectric/pull/2136/files

In particular how about at:-

https://github.com/jongerrish/robolectric/blob/018f55186d730fb654f099b21cd2b8af7c73cfc3/robolectric-resources/src/main/java/org/robolectric/res/ResType.java#L53 ?

**Comment by flyingtoaster0** [comment]
Cool, yeah that makes sense to do it there. Thanks for pointing that out. I'll give it a shot when I'm back at a keyboard

**Comment by flyingtoaster0** [comment]
The attributes don't seem to call through to ResType.getValueWithType() in the cases that were described in the issue, but I was still able to move the trim one layer deeper to ResName's constructor.

**Comment by jongerrish** [comment]
I was thinking more in the xml resource processing like maybe here:-

https://github.com/robolectric/robolectric/blob/9867a2f60d696d35f791df2a2ecdcf0f0f6ab64c/robolectric-resources/src/main/java/org/robolectric/res/StyleResourceLoader.java#L32

This seems to be where Attributes are created in this case.

I'd prefer if the processing of the xml was in one place like the parser, rather than the data objects personally.

You need to merge anyway, so do you want to take a look and let me know if you think this approach is feasible. If not I'm happy to merge either way.

**Comment by flyingtoaster0** [comment]
So I checked it out a bit more thoroughly, and it seems for resource names, the ResName object ultimately gets its ID from a MergedResourceIndex when then checks its sub-indexes for the presence of that ID in a HashMap<ResName, Integer>. In this case, we're looking at an OverlayResourceIndex. At the following line:

https://github.com/robolectric/robolectric/blob/9867a2f60d696d35f791df2a2ecdcf0f0f6ab64c/robolectric-resources/src/main/java/org/robolectric/res/OverlayResourceIndex.java#L53

we're grabbing that ID from the HashMap using a ResName that we're getting from ResName.withPackageName(). The problem here is that the "type" field that we're trying to trim is never used directly in order to access resource IDs; it's always contained within the ResName object itself. 

Without any major architectural changes, I feel like there are two simple options:

1. Do the trim in ResName's constructor
2. Do the trim in ResName.withPackageName()

Although both are still doing the trim inside ResName, option 2 at least gets it out of the constructor.

Let me know if you have any preferences, or if I'm still missing something

File: robolectric-resources/src/main/java/org/robolectric/res/ResName.java [hunk 1]
```diff
@@ -32,7 +32,7 @@ public ResName(@NotNull String fullyQualifiedName) {
       throw new IllegalStateException("\"" + fullyQualifiedName + "\" is not fully qualified");
     }
     packageName = matcher.group(NAMESPACE);
-    type = matcher.group(TYPE);
```
Comment (jongerrish) [inline robolectric-resources/src/main/java/org/robolectric/res/ResName.java]: Couldn't this constructor just call 

this(packageName, type, nameStr)

and save the duplication?

**Comment by jongerrish** [comment]
Thanks for the detailed explanation. I think the best thing is to take this as is then. If you wouldn't mind refactoring the constructor as mentioned in previous comment, I'll go ahead and merge it.

File: robolectric-resources/src/main/java/org/robolectric/res/ResName.java (hunk 1 above)
Comment (flyingtoaster0) [inline robolectric-resources/src/main/java/org/robolectric/res/ResName.java]: Calls to other constructors must be the first line in a constructor, but we first need that Matcher in order to get extract the packageName, type, and name. The next closest thing I can think of is having both constructors call some setup(String, String, String) method that just sets the values of the fields, but I feel like that gets to be a little overkill.

**Pull request merged**